server/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodes.java

```diff
@@ -448,7 +452,7 @@ public boolean hasChanges() {
         }
 
         public boolean masterNodeChanged() {
-            return newMasterNode != null;
+            return newMasterNode != previousMasterNode;
```

Comment:
This should use ephimeral ids, not object identity

Reply:
Right, it's a bit scary, sorry.